### PR TITLE
Refactor runner arg parsing for mulit-entry and multi-line

### PR DESCRIPTION
**Pull request by tomncooper**
This PR changes the handling of the args supplied to the SQLRunner main method so that it will:
* Allow SQL statements to be passed as a list (not one single long string)
* Allow each SQL statement passed to be multi-line and tabbed for easy readability
* Change the Statement Set handling so that sets remain in order and are not moved to the end

This means that instead of a using:

```yaml
  job:
    jarURI: local:///opt/streamshub/flink-sql-runner.jar
    args: ["Long string with multiple SQL statements in;"]
```
 we can now do:
```yaml
  job:
    jarURI: local:///opt/streamshub/flink-sql-runner.jar
    args:
      - |
        CREATE TABLE orders (
          order_number BIGINT, 
          price DECIMAL(32,2), 
          buyer ROW<first_name STRING, last_name STRING>, 
          last_name STRING, order_time TIMESTAMP(3)
        ) WITH (
          'connector' = 'datagen'
        );
      - |
        CREATE TABLE print_table 
        WITH (
          'connector' = 'print'
        ) LIKE orders;
      - INSERT INTO print_table SELECT * FROM orders;
 ```

This is a backwards compatible change and the long string array method will still work. 

This PR also changes the handling of [Statement Sets](https://docs.confluent.io/cloud/current/flink/reference/queries/statement-set.html) to match the example from the upstream Flink Kubernetes Operator example. Previously we would assume only 1 such statement set would be supplied and we would move it to the end of the statements list. It is possible there could be multiple statement sets and that later queries depend on the results created by them, so this PR changes the processing so that they remain in order.

This parsing is likely to need to be checked using more comprehensive system tests in [streams-e2e](https://github.com/streamshub/streams-e2e) which I will add once this PR has been reviewed.

### CODE REVIEW

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java [hunk 1]
```diff
@@ -74,44 +84,101 @@ public static void main(String[] args) throws Exception {
...
+        if (statement.startsWith("'") || statement.startsWith("\"")) {
+           statement = statement.substring(1);
+        }
+
+        if (statement.endsWith("'") || statement.endsWith("\"")) {
+            statement = statement.substring(0, statement.length() - 1);
+        }
+
+        return statement
+                .replaceAll("\n", " ")
```
Comment (kornys) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:172]:
```java
..replaceAll("\\R+", " ")
```

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java [hunk 2]
```diff
@@ -36,25 +36,35 @@ public class SqlRunner {
     private static final String STATEMENT_DELIMITER = ";"; // a statement should end with `;`
     private static final String LINE_DELIMITER = "\n";
 
+    private static final Pattern STATEMENT_DELIMETER_PATTERN = Pattern.compile("(?<!\\\\)" + STATEMENT_DELIMITER);
+
     private static final Pattern SET_STATEMENT_PATTERN =
-            Pattern.compile("SET\\s+'(\\S+)'\\s+=\\s+'(.*)';", Pattern.CASE_INSENSITIVE);
+            Pattern.compile("SET\\s+'(\\S+)'\\s*=\\s*'(.*)'\\s*;", Pattern.CASE_INSENSITIVE);
```
Comment (SamBarker) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:42]: Not really something for this PR but: at what point to do we decide Regex is N+1 problems and switch to a proper SQL parser (like Calcite)?

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java [hunk 3]
```diff
@@ -36,25 +36,35 @@ public class SqlRunner {
...
-    private static final Pattern STATEMENT_SET_PATTERN =
-            Pattern.compile("(EXECUTE STATEMENT SET BEGIN.*?END;)", Pattern.CASE_INSENSITIVE | Pattern.DOTALL);
+    private static final Pattern STATEMENT_SET_END_PATTERN = Pattern.compile("\\s*END\\s*;", Pattern.CASE_INSENSITIVE);
 
     public static void main(String[] args) throws Exception {
-        if (args.length != 1) {
-            throw new Exception("Exactly 1 argument is expected.");
+
+        System.out.println("\n####################################\n");
+        System.out.println("THESE ARE THE ARGS I HAVE BEEN SENT:");
```
Comment (SamBarker) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:52]: Logger at debug...

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java [hunk 4]
```diff
@@ -74,44 +84,101 @@ public static void main(String[] args) throws Exception {
...
+        // Split the statements on `;` except where that `;` is preceded by a double backspace
+        for (String statement: STATEMENT_DELIMETER_PATTERN.split(formatted)) {
+
+            if (!statement.isBlank()) {
+                // As we split on `;` we need to replace them at the end of the statement.
+                // There is probably a regex incantation we can add to the STATEMENT_DELIMETER_PATTERN to do this, but this will do for now.
+                statement = statement.trim() + STATEMENT_DELIMITER;
+                // Deal with the specific situation where secret strings in WITH clauses contain the `\\;` literal and will end up leaving a `\` before the `;`
+                // Again, there is probably regex foo to achieve this
+                statement = statement.replaceAll(Pattern.quote("\\;"), ";");
```
Comment (SamBarker) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:107]: nit: extract `Pattern.quote("\\;")` as a constant.

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java [hunk 5]
```diff
@@ -74,44 +84,101 @@ public static void main(String[] args) throws Exception {
...
-                current.append(c);
-                statements.add(current.toString().trim());
-                current = new StringBuilder();
+            var statementSetStartMatch = STATEMENT_SET_START_PATTERN.matcher(statement);
+
+            if (statementSetStartMatch.find()) {
+                LOG.debug("Found start of statement set: <begin>{}<end>", statement);
+                insideStatementSet = true;
+                statementSetString.append(statement);
+            } else if (insideStatementSet) {
```
Comment (SamBarker) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:119]: Hmm took me a bit of thinking to realise these two blocks can't collapse into one (given `insideStatementSet` is true when `statementSetStartMatch.find()` is true) because the become true on sequential evaluations of the loop. 

I get its logically the same thing (and sonar might whine) but it might be better to finish the `if` block with a `continue`.

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java [hunk 6]
```diff
@@ -74,44 +84,101 @@ public static void main(String[] args) throws Exception {
...
+                LOG.debug("Found statement inside statement set: <begin>{}<end>", statement);
+                statementSetString.append(" ").append(statement);
+                var statementSetEndMatch = STATEMENT_SET_END_PATTERN.matcher(statement);
+                if (statementSetEndMatch.find()) {
+                    LOG.debug("Found end of statement set: <begin>{}<end>", statement);
```
Comment (SamBarker) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:124]: I think it would be good to log the `statementSetString` we are appending here.

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java [hunk 7]
```diff
@@ -74,44 +84,101 @@ public static void main(String[] args) throws Exception {
...
+        var statements = new ArrayList<String>();
 
-        if (matcher.find()) {
-            statementSet = matcher.group(1);
-            otherStatements = formatted.replace(statementSet, "").trim();
-        }
+        var cleaned = cleanSqlStatement(rawStatementArg);
+        var formatted = formatSqlStatements(cleaned);
+
+        var statementSetString = new StringBuilder();
```
Comment (SamBarker) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:96]: `setStatmentBuilder`? `processedSetStatment` ?

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java [hunk 8]
```diff
@@ -74,44 +84,101 @@ public static void main(String[] args) throws Exception {
         }
     }
 
-    static List<String> parseStatements(String rawStatements) {
-        var formatted = formatSqlStatements(rawStatements.trim());
+    static List<String> parseStatementArg(String rawStatementArg) {
```
Comment (SamBarker) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:87]: Why re-name this method? `parseStatement` still feels like the right name.

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java [hunk 9]
```diff
@@ -35,20 +39,20 @@ void testParseStatementsTrailingSemiColon() {
...
-                "INSERT INTO print_table SELECT * FROM Table1; ";
+    void testParseStatementArgsWithSecrets() {
+        String statements = "CREATE TABLE Table1 (\nmessage STRING\n) WITH (\n\t'connector' = 'kafka',\n" +
+                        "\t'topic' = 'test',\n" +
+                        "\t'properties.bootstrap.servers' = 'my-cluster-kafka-bootstrap.flink.svc:9093',\n" +
+                        "\t'properties.security.protocol' = 'SASL_PLAINTEXT',\n" +
+                        "\t'properties.sasl.mechanism' = 'PLAIN',\n" +
+                        "\t'properties.sasl.jaas.config' = 'org.apache.kafka.common.security.plain.PlainLoginModule required username=\"{{secret:flink/mysecret/username}}\" password=\"{{secret:flink/mysecret/password}}\"\\;');\n" +
+                "CREATE TABLE print_table ( message STRING ) WITH ('connector' = 'print');\n" +
+                "INSERT INTO print_table SELECT * FROM Table1;\n";
```
Comment (SamBarker) [inline flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:55]: The textbook case for a textblock....

/me gets his coat 🃏

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java [hunk 10]
```diff
@@ -60,38 +64,40 @@ void testParseStatementsWithSecrets () {
...
                 "CREATE TABLE print_table ( message STRING ) WITH ('connector' = 'print');",
                 "CREATE TABLE print_table2 ( message STRING ) WITH ('connector' = 'print');",
                 "EXECUTE STATEMENT SET BEGIN INSERT INTO print_table SELECT * FROM KafkaTable; " +
-                "INSERT INTO print_table2 SELECT * FROM print_table; END;");
+                "INSERT INTO print_table2 SELECT * FROM print_table; END;",
+                "SELECT * FROM print_table2;");
 
-        List<String> actual = SqlRunner.parseStatements(statementSet);
+        List<String> actual = SqlRunner.parseStatementArgs(new String[]{statementSet});
         assertEquals(expected, actual);
```
Comment (SamBarker) [inline flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:101]: The PR description says the handling of `SET` statements has changed so they no longer sort to the end but this test hasn't change the verification of order. So I'm confused 😕 

I think the assertion is dependent on the equals implementation of the returned list. So maybe we should use [AssertJ#containsExactly](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/ListAssert.html#containsExactly-ELEMENT...-)

**Review by SamBarker — APPROVED** [review]
It certainly makes the inline args option more readable but thinking forward to loading statements from a config map or other external resource should we be doing something to support that.

Actually no lets merge this option and think about config maps separately.

Mostly nits/questions. Feel free to merge without waiting for me to re-review.

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java [hunk 11]
```diff
@@ -74,44 +84,101 @@ public static void main(String[] args) throws Exception {
...
-            statementSet = matcher.group(1);
-            otherStatements = formatted.replace(statementSet, "").trim();
-        }
+        var cleaned = cleanSqlStatement(rawStatementArg);
+        var formatted = formatSqlStatements(cleaned);
+
+        var statementSetString = new StringBuilder();
+        boolean insideStatementSet = false;
+        // Split the statements on `;` except where that `;` is preceded by a double backspace
+        for (String statement: STATEMENT_DELIMETER_PATTERN.split(formatted)) {
```
Comment (tombentley) [inline flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:99]: This is very broken. It's not enough to worry about `\\` quoting when SQL supports things like comments and string literals which could contain `;`. This cannot be done correctly using regex; a proper parser is needed. We should seriously look at reusing Calcite IMHO (since that's what Flink uses anyway, AFAIU).

**Review by tombentley — COMMENTED** [review]
I appreciate that this is a _kind of_ an improvement on what was there before, but I'm concerned that it's really just another step down a path that ultimately leads to lots of weird bugs for users.

**Comment by tomncooper** [comment]
Thanks for the review folks. I agree that, while this is a slightly better approach than the one we have currently, it is still too brittle and will be brought down by edge cases. 

I would like to avoid implementing our own parser logic, so I will have a look and see if we can use Flink's own parser logic to do the separation of SQL statements for us.

**Comment by tomncooper** [comment]
I'm gonna close this and come back with a more robust solution.

**Comment by SamBarker** [comment]
> I'm gonna close this and come back with a more robust solution.

I think that's letting perfect be the enemy of good enough. I'd rather me merged this as a step forward (not least because it makes any future demos that bit easier to run).

**Comment by tomncooper** [comment]
Ok, let me address the issues in this and in the mean time I will look at the Flink parser

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java (hunk 3 above)
Comment (tomncooper) [reply flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:52]: Well that definitely shouldn't still be in there. You are seeing the edges of my madness, trying to figure out what K8s was actually passing to the main method...

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java (hunk 2 above)
Comment (tomncooper) [reply flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:42]: Yeah, agreed. I will look at using the Flink SQL parser logic.

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java (hunk 8 above)
Comment (tomncooper) [reply flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:87]: Well now we are parsing all the args, which each could be just one statement or multiple, it felt better to call a spade a spade.

File: flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java (hunk 10 above)
Comment (tomncooper) [reply flink-sql-runner/src/test/java/com/github/streamshub/flink/SqlRunnerTest.java:101]: The order of "statement sets" (not `SET` commands) in the statement list is now preserved, where as previously we moved the (assumed to be only 1) statement set to the end. 

I added a statement _after_ the statement set to make sure the order wasn't changed.

**Comment by kornys** [comment]
> Ok, let me address the issues in this and in the mean time I will look at the Flink parser

and also please rebase it to latest main :)

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java (hunk 7 above)
Comment (tomncooper) [reply flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:96]: Changed it to option A

File: flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java (hunk 5 above)
Comment (tomncooper) [reply flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java:119]: I added an else block and some more descriptive comments to hopefully make it clear what is going on.

**Comment by kornys** [comment]
@tomncooper good news, backward compatibility works :D

**Comment by katheris** [comment]
From what I can see a fair amount of the complex logic here centers around having the option to pass multiple statements per arg when you provide multiple arguments. Would it be simpler to have this PR allow you to pass statements as a list, but assume that if the list is of size > 1 each list entry is a single statement? Then use the Flink parser in the next PR to allow multiple statements within each arg in a list of args.

That way you can provide the statements in a more user friendly way but also keeps the codebase simpler. At the moment it's quite hard to follow.

**Comment by tomncooper** [comment]
@katheris I agree a staged approach here would be better. I have done more testing and I now think this is not the way to go. I am going to close this PR and come back with a simpler implementation that just address the multi-line strings. We can then move to a more robust parser.
